extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/K8sTaskAdapter.java

```diff
@@ -0,0 +1,277 @@
+ * to add some extra coordination to shut down sidecar containers when the main pod exits.
+ */
+
+public abstract class K8sTaskAdapter implements TaskAdapter<Pod, Job>
+{
+
+  private static final EmittingLogger log = new EmittingLogger(K8sTaskAdapter.class);
+
+  protected final KubernetesClientApi client;
+  protected final KubernetesTaskRunnerConfig config;
+  protected final ObjectMapper mapper;
+
+  public K8sTaskAdapter(
+      KubernetesClientApi client,
+      KubernetesTaskRunnerConfig config,
+      ObjectMapper mapper
+  )
+  {
+    this.client = client;
+    this.config = config;
+    this.mapper = mapper;
+  }
+
+  @Override
+  public Job fromTask(Task task, PeonCommandContext context) throws IOException
+  {
+    String myPodName = System.getenv("HOSTNAME");
+    Pod pod = client.executeRequest(client -> client.pods().inNamespace(config.namespace).withName(myPodName).get());
+    return createJobFromPodSpec(pod.getSpec(), task, context);
+  }
+
+  @Override
+  public Task toTask(Pod from) throws IOException
+  {
+    List<EnvVar> envVars = from.getSpec().getContainers().get(0).getEnv();
+    Optional<EnvVar> taskJson = envVars.stream().filter(x -> "TASK_JSON".equals(x.getName())).findFirst();
+    String contents = taskJson.map(envVar -> taskJson.get().getValue()).orElse(null);
+    if (contents == null) {
+      throw new IOException("No TASK_JSON environment variable found in pod: " + from.getMetadata().getName());
+    }
+    return mapper.readValue(Base64Compression.decompressBase64(contents), Task.class);
+  }
+
+  @VisibleForTesting
+  public abstract Job createJobFromPodSpec(PodSpec podSpec, Task task, PeonCommandContext context) throws IOException;
+
+  protected Job buildJob(
+      K8sTaskId k8sTaskId,
+      Map<String, String> labels,
+      Map<String, String> annotations, PodTemplateSpec podTemplate
+  )
+  {
+    return new JobBuilder()
+        .withNewMetadata()
+        .withName(k8sTaskId.getK8sTaskId())
+        .addToLabels(labels)
+        .addToAnnotations(annotations)
+        .endMetadata()
+        .withNewSpec()
+        .withTemplate(podTemplate)
+        .withActiveDeadlineSeconds(config.maxTaskDuration.toStandardDuration().getStandardSeconds())
+        .withBackoffLimit(0)
+        .withTtlSecondsAfterFinished((int) config.taskCleanupDelay.toStandardDuration().getStandardSeconds())
+        .endSpec()
+        .build();
+  }
+
+  @VisibleForTesting
+  static Optional<Long> getJavaOptValueBytes(String qualifier, List<String> commands)
+  {
+    Long result = null;
+    Optional<String> lastHeapValue = commands.stream().filter(x -> x.startsWith(qualifier)).reduce((x, y) -> y);
+    if (lastHeapValue.isPresent()) {
+      result = HumanReadableBytes.parse(StringUtils.removeStart(lastHeapValue.get(), qualifier));
+    }
+    return Optional.ofNullable(result);
+  }
+
+  @VisibleForTesting
+  static long getContainerMemory(PeonCommandContext context)
```

Comment:
Comments might be useful for this method on how the container memory is estimated.

Reply:
good call, will add it